formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/Helpers.kt

```diff
@@ -65,22 +71,64 @@ internal fun SerialDescriptor.extractParameters(index: Int): ProtoDesc {
             format = annotation.type
         } else if (annotation is ProtoPacked) {
             protoPacked = true
+        } else if (annotation is ProtoOneOf) {
+            isOneOf = true
         }
     }
-    return ProtoDesc(protoId, format, protoPacked)
+    if (isOneOf) {
+        // reset proto to index-based for decoding,
+        // proto id annotated in oneOf field has no meaning
+        protoId = index + 1
+    }
+    return ProtoDesc(protoId, format, protoPacked, isOneOf)
 }
 
+internal fun SerialDescriptor.extractClassDesc(): ProtoDesc {
+    var protoId: Int = -1
+    var isOneOf = false
+    for (i in annotations.indices) {
+        val annotation = annotations[i]
+        if (annotation is ProtoNumber) {
+            protoId = annotation.number
+            isOneOf = true
+        }
+    }
+    return if (protoId == -1) {
+        MISSING_TAG
+    } else {
+        ProtoDesc(protoId, ProtoIntegerType.DEFAULT, oneOf = isOneOf)
+    }
+}
+
+/**
+ * Get the proto id from the descriptor of [index] element,
+ * or return [ID_HOLDER_ONE_OF] if such element is marked with [ProtoOneOf]
+ */
 internal fun extractProtoId(descriptor: SerialDescriptor, index: Int, zeroBasedDefault: Boolean): Int {
     val annotations = descriptor.getElementAnnotations(index)
     for (i in annotations.indices) { // Allocation-friendly loop
         val annotation = annotations[i]
-        if (annotation is ProtoNumber) {
+        if (annotation is ProtoOneOf) {
+            return ID_HOLDER_ONE_OF
+        } else if (annotation is ProtoNumber) {
             return annotation.number
         }
     }
     return if (zeroBasedDefault) index else index + 1
 }
 
+private val emptyArray by lazy(mode = LazyThreadSafetyMode.NONE) { IntArray(0) }
```

Comment:
This should be a constant (I don't remember whether `emptyIntArray` actually exists), not an instance final value. At that point it makes little sense to have the overhead of making it lazy.

Reply:
Sadly there is no builtin `emptyIntArray` in kotlin, but I found that I can just return null if no annotation here.